### PR TITLE
Reserve memory only when needed (mark 2)

### PR DESCRIPTION
**Pull request by davidgiven**
See: https://github.com/amery/linux-allwinner/issues/18

This is quite a lot cleaner; using #if...#endif is unavoidable, I'm afraid, but now there's only one block per case and it's much better laid out.

Because I think the pull request is going to show the various commits that make it up separately, which makes it a bit hard to follow, the overall diff against upstream is here: 

https://github.com/davidgiven/linux-allwinner/compare/amery:allwinner-v3.0-android-v2...davidgiven:dg-memory

### CODE REVIEW

**Comment by hno** [comment]
github kindly provides diffs in pull requests both in the mail and online (diff tab)

loojks good to me.

**Comment by hno** [comment]
should also be possible to return the last 64MB reserved for mali. But thats a separate topic.

**Comment by davidgiven** [comment]
Awesome. Thanks.
